### ISSUE TITLE
Function return type with variadic tuple does not allow empty tuple

### ISSUE THREAD

**Issue by AlCalzone** [microsoft/TypeScript#40491]
**TypeScript Version:**  4.1.0-insiders.20200910

**Search Terms:** Variadic tuple empty return

**Code**

```ts
function foo<T extends [] = []>(): [...T, string] {
  return [""];
  // Error: Source has 1 element, but target requires 2
}

const bar = foo();
//    ^ has expected type [string];
```

**Expected behavior:** No error

**Actual behavior:** Error on the return statement. Even if I change the constraint to `extends any[]`, it does not work.

**Playground Link:** https://www.typescriptlang.org/play?ts=4.1.0-insiders.20200910#code/GYVwdgxgLglg9mABMOcA8AVRBTAHlbMAEwGdEBtAXUQF4LKA+ACgEoAuCgOm4wBpESUAE4wwAc2oBvAFCJEQ7FBBCk5AERrKAblmIA9HsQBRIULhCOAZTjKI2RAAsAhmQCMOADbYAtoSj8AIxAoRCgnITFFeWwARxAYBTIAJmkAX2lpCARBRADw2mRUVh0DOTkAPUcXHFwAB2xobCJQgE96ikERcW0gA

**Related Issues:** A bunch related to variadic tuples, but none seemed to match this case

**Comment by AlCalzone** [comment]
@ahejlsberg

**Comment by RyanCavanaugh** [comment]
For `extends any[]`, this is the correct behavior:
```ts
function foo<T extends any[]>(): [...T, string] {
  // Let's assume this works?
  return [""] as any;
}

const bar = foo<[1, 2]>();
// Crash
bar[0].toFixed();
```

For `extends tuple_like`, this works as long as you actually spread in an appropriate value:
```ts
// All correct implementations of 'foo' look like this
function foo<T extends [unknown]>(arg: T): [...T, string] {
  return [...arg, ""];
}

// Works
const bar = foo([1]);
```

For the *specific case* of `extends []`, what's being suggested here then is that we specifically special-case a zero-length tuple to allow it to be skipped in the output position, but it really doesn't seem like there's any reason whatsoever to write a constraint like except as an accident, so I'm calling this working as intended.

**Comment by AlCalzone** [comment]
@RyanCavanaugh okay, let's give it something to spread:

```ts
function foo<T extends any[]>(arg: T): [...T, string] {
  return [...arg, ""];
}

const test1 = foo([]);
//    ^ has type string[], expected [string]
console.log(JSON.stringify(test1)); // [""]

const test2 = foo([1]);
//    ^ has type (string | number)[], expected [number, string]
console.log(JSON.stringify(test2)); // [1, ""]
```

Maybe I'm missing the point, but shouldn't I be able to say "this function returns a tuple where the last element is a string"?

**Comment by RyanCavanaugh** [comment]
```ts
function foo<T extends readonly any[]>(arg: T): [...T, string] {
  return [...arg, ""];
}

const test1 = foo([] as const);
//    ^ has type [string]
console.log(JSON.stringify(test1)); // [""]

const test2 = foo([1] as const);
//    ^ has type [number, string]
console.log(JSON.stringify(test2)); // [1, ""]
```

**Comment by orta** [comment]
@AlCalzone  - how did you make that playground URL? all the playground URLS have `X.Y.Z-dev` not `X.Y.Z-internal`

![Screen Shot 2020-09-11 at 12 27 10 PM](https://user-images.githubusercontent.com/49038/92950009-fe916780-f429-11ea-92e1-c97c94b58369.png)

**Comment by AlCalzone** [comment]
> @AlCalzone  - how did you make that playground URL?

I was confused too, but I just clicked 'next'.

@RyanCavanaugh Thanks
